metrotranscom/doorway
# feat: ses email setup

# Pull Request Template

## Issue Overview

This PR addresses [323](https://github.com/metrotranscom/doorway/issues/323)


## Description

Notes

* The reason that we're using nodemailer is because of our potential need to send emails with attachments. Out of the box, the SES sendEmail function does everything we need it to do, but they don't have a nice way to add attachments to emails. To add an attachment, you need to use sendRawEmail, which makes you build a MIME formatted email from scratch, which is icky - nodemailer wraps the SES API and allows us to do that more easily. Even though we only have one potential need for the CSV export attachment, I'm using nodemailer everywhere for consistency.
* SES has a system for managing templates and using templated data in handlebars files which is what I tried initially for our emails, but there isn't a great system for creating / updating / managing the templates in SES. You need to hit the API directly, and there isn't a good way of managing them in the AWS UI. It would be somewhat of a pain to need to store handlebars files, convert them to strings, and send em to SES whenever we want to make a string change. I'm continuing the same pattern we were using before, which is that we just smush the data we need into an already compiled handlebars template and send all of that as the email HTML, without actually using templated data. While templates may offer some nice benefits, an additional problem was that if you have any malformatted or missing template data, SES will not send the email, but also intentionally won't send you back a failure in the API - it will silently fail. The failure indication would only be available in CloudWatch logs which is a bummer.
* There is no longer a need in Doorway to have an emailFromAddress as everything will be no-reply@housingbayarea.org
* Logs for our SES account are manually routed to a dashboard in CloudWatch. There are additional tags we can add to emails for CloudWatch if we want to be able to identify more details about certain email metrics. It's kind of hard to tell without more data which forms of displaying the data are going to be the most useful. Understanding which tags we want might be iterative as we have different queries of log data. At the moment we have data on sends / fails.

## How Can This Be Tested/Reviewed?

Send emails locally and ensure they come through and are formatted correctly.

Remaining questions
* Do we want the same retry functionality we had for SendGrid?
* Do we want to use nodemailer's built-in rate limiting?

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
